File: api/transcript.py
from http.server import BaseHTTPRequestHandler
from youtube_transcript_api import YouTubeTranscriptApi
from youtube_transcript_api.formatters import TextFormatter
import json

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()
        
    def do_POST(self):
        try:
            # Set CORS headers
            self.send_response(200)
            self.send_header('Access-Control-Allow-Origin', '*')
            self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
            self.send_header('Access-Control-Allow-Headers', 'Content-Type')
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            
            # Read request body
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length)
            data = json.loads(post_data.decode('utf-8'))
            
            video_id = data.get('videoId')
            if not video_id:
                response = json.dumps({'error': 'Missing videoId parameter'})
                self.wfile.write(response.encode('utf-8'))
                return
            
            # Fetch transcript using youtube-transcript-api
            try:
                # Try to get transcript in English first
                transcript_list = YouTubeTranscriptApi.list_transcripts(video_id)
                
                # Look for English transcript
                transcript = None
                language_code = 'en'
                is_generated = False
                
                try:
                    # Try manually created English transcript first
                    transcript = transcript_list.find_transcript(['en'])
                    language_code = 'en'
                    is_generated = False
                except:
                    try:
                        # Try generated English transcript
                        transcript = transcript_list.find_generated_transcript(['en'])
                        language_code = 'en'
                        is_generated = True
                    except:
                        # Fall back to any available transcript
                        available_transcripts = list(transcript_list)
                        if available_transcripts:
                            transcript = available_transcripts[0]
                            language_code = transcript.language_code
                            is_generated = transcript.is_generated
                        else:
                            response = json.dumps({
                                'error': 'No transcript available for this video'
                            })
                            self.wfile.write(response.encode('utf-8'))
                            return
                
                # Fetch the actual transcript data
                transcript_data = transcript.fetch()
                
                # Format transcript as plain text
                formatter = TextFormatter()
                text_transcript = formatter.format_transcript(transcript_data)
                
                # Also provide structured data with timestamps
                structured_transcript = []
                for entry in transcript_data:
                    structured_transcript.append({
                        'text': entry['text'],
                        'start': entry['start'],
                        'duration': entry['duration']
                    })
                
                response_data = {
                    'transcript': text_transcript,
                    'structured_transcript': structured_transcript,
                    'language_code': language_code,
                    'is_generated': is_generated,
                    'total_entries': len(transcript_data)
                }
                
                response = json.dumps(response_data)
                self.wfile.write(response.encode('utf-8'))
                
            except Exception as transcript_error:
                error_message = str(transcript_error)
                if 'TranscriptsDisabled' in error_message:
                    error_message = 'Transcripts are disabled for this video'
                elif 'NoTranscriptFound' in error_message:
                    error_message = 'No transcript available for this video'
                elif 'VideoUnavailable' in error_message:
                    error_message = 'Video is unavailable or private'
                
                response = json.dumps({'error': error_message})
                self.wfile.write(response.encode('utf-8'))
                
        except Exception as e:
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            response = json.dumps({
                'error': 'Internal server error',
                'details': str(e)
            })
            self.wfile.write(response.encode('utf-8'))
            
    def do_GET(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        response = json.dumps({'message': 'Transcript API is running'})
        self.wfile.write(response.encode('utf-8'))

File: api/transcript.js
module.exports = async function handler(req, res) {
  // Dynamic import for ES module
  const { fetchTranscript } = await import('youtube-transcript-plus');
  // CORS: reflect origin and handle preflight robustly
  const origin = req.headers.origin || '*';
  res.setHeader('Vary', 'Origin');
  if (origin && origin !== 'null') {
    res.setHeader('Access-Control-Allow-Origin', origin);
  } else {
    res.setHeader('Access-Control-Allow-Origin', '*');
  }
  res.setHeader('Access-Control-Allow-Credentials', 'true');
  res.setHeader('Access-Control-Allow-Methods', 'GET,OPTIONS,PATCH,DELETE,POST,PUT');
  const acrh = req.headers['access-control-request-headers'];
  res.setHeader('Access-Control-Allow-Headers', acrh || 'Content-Type, Authorization, X-Requested-With');
  res.setHeader('Access-Control-Max-Age', '86400');

  // Handle preflight requests
  if (req.method === 'OPTIONS') {
    res.status(204).end();
    return;
  }

  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  try {
    const { videoId, config = {} } = req.body;
    
    if (!videoId) {
      return res.status(400).json({ error: 'Missing videoId parameter' });
    }



    // Default configuration - prefer English, use a browser-like user agent
    const defaultConfig = {
      lang: 'en',
      userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36',
      ...config
    };

    try {
      // Use youtube-transcript-plus to fetch transcript
      const transcriptArray = await fetchTranscript(videoId, defaultConfig);

      if (!transcriptArray || transcriptArray.length === 0) {
        return res.status(404).json({ 
          error: 'No transcript data found for this video',
          videoId: videoId 
        });
      }

      // Format the response similar to the original API structure
      const formattedTranscript = transcriptArray.map(item => ({
        text: item.text,
        start: item.offset,
        duration: item.duration
      }));

      // Create plain text version
      const plainText = transcriptArray.map(item => item.text).join(' ');

      const responseData = {
        success: true,
        videoId: videoId,
        transcript: formattedTranscript,
        plainText: plainText,
        metadata: {
          language: transcriptArray[0]?.lang || defaultConfig.lang,
          isGenerated: true, // youtube-transcript-plus typically gets auto-generated transcripts
          segmentCount: transcriptArray.length,
          totalLength: plainText.length,
          duration: transcriptArray.reduce((total, item) => total + (item.duration || 0), 0)
        }
      };



      return res.status(200).json(responseData);

    } catch (transcriptError) {
      console.error('youtube-transcript-plus error:', transcriptError);
      
      // Handle specific error types from youtube-transcript-plus
      let errorMessage = 'Failed to fetch transcript';
      let statusCode = 500;

      if (transcriptError.name === 'YoutubeTranscriptVideoUnavailableError') {
        errorMessage = 'Video is unavailable or has been removed';
        statusCode = 404;
      } else if (transcriptError.name === 'YoutubeTranscriptDisabledError') {
        errorMessage = 'Transcripts are disabled for this video';
        statusCode = 403;
      } else if (transcriptError.name === 'YoutubeTranscriptNotAvailableError') {
        errorMessage = 'No transcript is available for this video';
        statusCode = 404;
      } else if (transcriptError.name === 'YoutubeTranscriptNotAvailableLanguageError') {
        errorMessage = `Transcript is not available in the requested language: ${defaultConfig.lang}`;
        statusCode = 404;
      } else if (transcriptError.name === 'YoutubeTranscriptTooManyRequestError') {
        errorMessage = 'Too many requests. Please try again later';
        statusCode = 429;
      } else if (transcriptError.name === 'YoutubeTranscriptInvalidVideoIdError') {
        errorMessage = 'Invalid video ID or URL provided';
        statusCode = 400;
      }

      return res.status(statusCode).json({ 
        error: errorMessage,
        videoId: videoId,
        errorType: transcriptError.name || 'UnknownError',
        details: transcriptError.message
      });
    }
    
  } catch (error) {
    console.error('API Error:', error);
    return res.status(500).json({ 
      error: 'Internal server error',
      details: error.message
    });
  }
}
